analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CloudNative/AzureFunctionsLogFailures.cs

```diff
@@ -0,0 +1,154 @@
+﻿using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Azure.WebJobs;
+using Microsoft.Azure.WebJobs.Extensions.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using System;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace AzureFunctions1
+{
+    public static class Function1
+    {
+        [FunctionName("Function1")]
+        public static async Task<IActionResult> EmptyCatchClause([HttpTrigger(AuthorizationLevel.Function, "get", "post", Route = null)] HttpRequest req, ILogger log)
+        {
+            try
+            {
+                return new EmptyResult();
+            }
+            catch // Noncompliant {{Log caught exceptions via ILogger with LogLevel Warning, Error, or Critical}}
+//          ^^^^^
+            {
+                return new EmptyResult();
+            }
+        }
+
+        [FunctionName("Function1")]
+        public static async Task<IActionResult> LogLevelInvalid([HttpTrigger(AuthorizationLevel.Function, "get", "post", Route = null)] HttpRequest req, ILogger log)
+        {
+            try
+            {
+                return new EmptyResult();
+            }
+            catch // Noncompliant
+//          ^^^^^
+            {
+                log.LogTrace(string.Empty);        // Secondary
+//              ^^^^^^^^^^^^^^^^^^^^^^^^^^
+                log.LogInformation(string.Empty);  // Secondary
+                log.LogDebug(string.Empty);        // Secondary
+                return new EmptyResult();
+            }
+        }
+
+        [FunctionName("Function1")]
+        public static async Task<IActionResult> LogExceptionInCatchClause([HttpTrigger(AuthorizationLevel.Function, "get", "post", Route = null)] HttpRequest req, ILogger log)
+        {
+            try
+            {
+                return new EmptyResult();
+            }
+            catch (Exception ex) // Compliant
+            {
+                log.LogError(ex, "");
+                return new EmptyResult();
+            }
+        }
+
+        [FunctionName("Function1")]
+        public static async Task<IActionResult> LogExceptionInWrappedLogger([HttpTrigger(AuthorizationLevel.Function, "get", "post", Route = null)] HttpRequest req, ILogger log)
+        {
+            try
+            {
+                return new EmptyResult();
+            }
+            catch (Exception ex) // Compliant
+            {
+                var nullLogger = NullLogger.Instance;
+                nullLogger.Log(LogLevel.Error, new EventId(), (object)null, ex, (s, e) => string.Empty);
+                return new EmptyResult();
+            }
+        }
+
+        private static bool True(Action action)
```

Comment:
_Low priority, our coding tyle doesn't apply in test cases_ I'd put this private helper and the end of the class. It's hard to find somewhere here in the middle.